### PR TITLE
SER-3907 Start storing private user properites in user-preferences

### PR DESCRIPTION
**Pull request by mmeller-wikia**
https://wikia-inc.atlassian.net/browse/SER-3907

### CODE REVIEW

File: includes/User.php [hunk 1]
```diff
@@ -2367,7 +2367,13 @@ public function getOption($oname, $defaultOverride = null, $ignoreHidden = false
 	 */
 	private function getOptionHelper( $oname, $defaultOverride = null, $ignoreHidden = false ) {
 		global $wgHiddenPrefs;
+		global $wgPrivateUserAttributes;
+
 		$this->loadOptions();
+		if ( in_array( $oname, $wgPrivateUserAttributes ) && array_key_exists( $oname, $this->mOptions ) ) {
+			$options = $this->getOptions();
+			return array_key_exists( $oname, $options ) ? $options[$oname] : $defaultOverride;
+		}
```
Comment (pwojcik86) [inline includes/User.php]: I don't understand why this change is needed. If we save a property to the db and `user-preferences`, just reading from the db should be fine. Could you explain this change?

File: includes/User.php [hunk 2]
```diff
@@ -4146,7 +4152,14 @@ protected function loadOptions() {
...
 	protected function savePreferences() {
-		$this->userPreferences()->save($this->getId());
+		$additionalGlobalOptions = [];
+		foreach( $this->mOptions as $key => $value ) {
+			if ( $this->shouldOptionBeStored( $key, $value ) ) {
+				$additionalGlobalOptions[$key] = $value;
+			}
+		}
+
+		$this->userPreferences()->save($this->getId(), $additionalGlobalOptions);
```
Comment (pwojcik86) [inline includes/User.php:4135]: Are `$additionalGlobalOptions` necessary if you've already added all "private" properties to `$wgGlobalUserPreferenceWhiteList`?

**Review by pwojcik86 — CHANGES_REQUESTED** [review]

File: includes/User.php (hunk 2 above)
Comment (mmeller-wikia) [reply includes/User.php:4135]: It's used to pass all needed private properties to save, in other case we will save only changed preferences and all read from user-preferences.

File: includes/User.php (hunk 1 above)
Comment (mmeller-wikia) [reply includes/User.php]: This one fixes saving options to user-preferences from Spcial:Preferences which were modified on ucp

Comment (mmeller-wikia) [reply includes/User.php]: Probably it is caching problem

**Review by pwojcik86 — APPROVED** [review]

**Review by pwojcik86 — APPROVED** [review]
